resources/assets/js/hooks/webResourceHooks/indexCrudReducer.ts

```diff
@@ -0,0 +1,415 @@
+  | IndexFulfillAction<T>
+  | IndexRejectAction
+  | CreateStartAction<T>
+  | CreateFulfillAction<T>
+  | CreateRejectAction<T>
+  | UpdateStartAction<T>
+  | UpdateFulfillAction<T>
+  | UpdateRejectAction<T>
+  | DeleteStartAction
+  | DeleteFulfillAction
+  | DeleteRejectAction;
+
+export interface ResourceState<T> {
+  indexMeta: {
+    status: ResourceStatus;
+    pendingCount: number;
+    error: Error | FetchError | undefined;
+  };
+  createMeta: {
+    status: ResourceStatus;
+    pendingCount: number;
+    error: Error | FetchError | undefined; // Only stores the most recent error;
+  };
+  values: {
+    [id: string]: {
+      value: T;
+      error: Error | FetchError | undefined;
+      status: ResourceStatus;
+      pendingCount: number;
+    };
+  };
+}
+
+export function initializeState<T extends { id: number }>(
+  items: T[],
+): ResourceState<T> {
+  return {
+    indexMeta: {
+      status: "initial",
+      pendingCount: 0,
+      error: undefined,
+    },
+    createMeta: {
+      status: "initial",
+      pendingCount: 0,
+      error: undefined,
+    },
+    values: mapToObjectTrans(items, getId, (item) => ({
+      value: item,
+      error: undefined,
+      status: "initial",
+      pendingCount: 0,
+    })),
+  };
+}
+
+type StateValues<T> = ResourceState<T>["values"];
+
+/**
+ * Decrement the number if it above zero, else return 0.
+ * @param num
+ */
+function decrement(num: number): number {
+  return num <= 0 ? 0 : num - 1;
+}
+
+function mergeItem<T extends { id: number }>(
+  values: StateValues<T>,
+  item: T,
+): StateValues<T> {
+  if (hasKey(values, item.id)) {
+    // We leave the status and count as is, in case an update or delete is in progress for this item.
+    return {
+      ...values,
+      [item.id]: {
+        ...values[item.id],
+        value: item,
+        error: undefined,
+      },
+    };
+  }
+  return {
+    ...values,
+    [item.id]: {
+      value: item,
+      error: undefined,
+      status: "fulfilled",
+      pendingCount: 0,
+    },
+  };
+}
+function mergeIndexPayload<T extends { id: number }>(
+  values: StateValues<T>,
+  payload: T[],
+): StateValues<T> {
+  // Update or create a values entry for each item in the payload.
+  const newValues = payload.reduce(mergeItem, values);
+  // Delete any values entries that don't exist in the new payload.
+  const payloadIds = payload.map(getId);
+  return filterObjectProps(newValues, (item) =>
+    payloadIds.includes(item.value.id),
+  );
+}
+
+function mergeCreatePayload<T extends { id: number }>(
+  values: StateValues<T>,
+  payload: T,
+): StateValues<T> {
+  if (hasKey(values, payload.id)) {
+    // Something has gone wrong if an existing item has the same id as the newly created one.
+    // TODO: But should we throw the error, or just update the value and pretend everything's ok?
+    throw new Error(
+      "Cannot create new item as an existing item shares the same id. Try refreshing the whole index.",
+    );
```

Comment:
From what I can tell, there's no way to stop the hook from crashing the entire component when it throws an error, so I've avoided doing so in the reducer. It may be reasonable to blindly trust whatever is received from the server anyway.